packages/picasso.js/src/core/geometry/geopolygon.js

```diff
@@ -0,0 +1,225 @@
+import { create as createPolygon } from './polygon';
+import {
+  pointsToLine,
+  pointsToRect
+} from './util';
+import {
+  testCircleGeoPolygon,
+  testGeoPolygonPoint,
+  testGeoPolygonLine,
+  testGeoPolygonRect,
+  testRectRect,
+  testRectContainsRect,
+  testPolygonLine
+} from '../math/narrow-phase-collision';
+
+function close(vertices) {
+  const first = vertices[0];
+  const last = vertices[vertices.length - 1];
+
+  if (first.x !== last.x || first.y !== last.y) {
+    vertices.push(first);
+  }
+}
+
+function removeDuplicates(vertices) {
+  for (let i = 0; i < vertices.length - 1; i++) {
+    const v0 = vertices[i];
+    const v1 = vertices[i + 1];
+    if (v0.x === v1.x && v0.y === v1.y) {
+      vertices.splice(i, 1);
+      i--;
+    }
+  }
+}
+
+function testGeoPolygonPolygon(geopolygon, polygon) {
+  let intersects = false;
+  for (let i = 0, len = polygon.edges.length; i < len; i++) {
+    intersects = testGeoPolygonLine(geopolygon, pointsToLine(polygon.edges[i]));
+    if (intersects === true) {
+      break;
+    }
+  }
+  return intersects;
+}
+
+function testPolygonGeoPolygon(polygon, geopolygon) {
+  let intersects = false;
+  const { numPolygons, polygons } = geopolygon;
+  for (let n = 0; n < numPolygons; n++) {
+    let polygon2 = polygons[n];
+    for (let i = 0, len = polygon2.edges.length; i < len; i++) {
+      intersects = testPolygonLine(polygon, pointsToLine(polygon2.edges[i]));
+      if (intersects === true) {
+        return true;
+      }
+    }
+    return false;
+  }
+  for (let i = 0, len = polygon.edges.length; i < len; i++) {
+    intersects = testGeoPolygonLine(geopolygon, pointsToLine(polygon.edges[i]));
+    if (intersects === true) {
+      break;
+    }
+  }
+  return intersects;
+}
+/**
+ * A geo-polygon is a polygon which is similar to a polygon in GeoJson. A typical geopolygon is an array of polygons where the first polygon is an outer polygon and the rest are inner polygons
+ * @private
+ */
+
+/**
+ * Construct a new GeoPolygon instance
+ * @private
+ */
+class GeoPolygon {
+  constructor({ vertices = [[]] } = {}) {
```

Comment:
I don't think naming the variable `vertices` make any sense. It's an array of "polygons" or an array of "polygon definitions". It might be hard to find a good name for it. I'd suggest the one of the following: ´data´ or ´polygons´

Reply:
This naming style respects GeoPolygon style..